arch/x86/core/crt0.S

```diff
@@ -15,6 +15,93 @@
 #include <kernel_arch_data.h>
 #include <arch/cpu.h>
 
+#ifdef CONFIG_JAILHOUSE
+
+#define JAILHOUSE_INMATE_CS32   0x8
+#define JAILHOUSE_INMATE_DS32   0x18
+#define JAILHOUSE_TEXT_OFFSET   0xf0000
+#define MSR_MTRR_DEF_TYPE       0x000002ff
+#define MTRR_ENABLE             0x00000800
+#define X86_CR0_PE              0x00000001
+#define X86_CR4_PSE             0x00000010
+
+	.code16
+	.section ".boot", "ax"
+
+	.globl __jh_entry
+__jh_entry:
+	ljmp $0xf000,$start16
+
+	.section ".startup", "ax"
+start16:
+	lgdtl %cs:gdt_ptr
+
+	mov %cr0,%eax
+	or $X86_CR0_PE,%al
+	mov %eax,%cr0
+
+	ljmpl $JAILHOUSE_INMATE_CS32,$start32 + JAILHOUSE_TEXT_OFFSET
+
+	.code32
+start32:
+	/*
+	 * Minimal bootstrap into 32-bit mode, just to jump to
+	 * __start
+	 */
+
+	/* 4Mb pages */
+	mov %cr4,%eax
+	or $X86_CR4_PSE,%eax
+	mov %eax,%cr4
+
+	mov $loader_pdpt,%eax
+	mov %eax,%cr3
+
+	/* Enable paging, write protect and protected mode */
```

Comment:
I'd like to learn more about why we are enabling paging here. Jailhouse requirement?

Reply:
I guess they wanted that completed at this boot phase there, I see we don't need this here. Thanks.